modules/fluid_properties/src/userobjects/CaloricImperfectGas.C

```diff
@@ -0,0 +1,970 @@
+
+CaloricImperfectGas::CaloricImperfectGas(const InputParameters & parameters)
+  : SinglePhaseFluidProperties(parameters),
+    NaNInterface(this),
+    _molar_mass(getParam<Real>("molar_mass")),
+    _R_specific(_R / _molar_mass),
+    _T_c(getParam<Real>("T_c")),
+    _rho_c(getParam<Real>("rho_c")),
+    _e_c(getParam<Real>("e_c")),
+    _min_temperature(getParam<Real>("min_temperature")),
+    _max_temperature(getParam<Real>("max_temperature")),
+    _tol(1e-4)
+{
+}
+
+void
+CaloricImperfectGas::initialSetup()
+{
+  SinglePhaseFluidProperties::initialSetup();
+  _h_T = &getFunction("h");
+  _e_T = &getFunction("e");
+  _cp_T = &getFunction("cp");
+  _cv_T = &getFunction("cv");
+  _mu_T = &getFunction("mu");
+  _k_T = &getFunction("k");
+
+  setupLookupTables();
+}
+
+CaloricImperfectGas::~CaloricImperfectGas() {}
+
+void
+CaloricImperfectGas::setupLookupTables()
+{
+  // ensure e(T) and h(T) are monotonic
+  unsigned int n = std::floor(_max_temperature - _min_temperature) + 1;
+  _delta_T = (_max_temperature - _min_temperature) / ((Real)n - 1.0);
+  Real e_old;
+  Real h_old;
+  for (unsigned int j = 0; j < n; ++j)
+  {
+    Real temperature = _min_temperature + j * _delta_T;
+    Real h = _h_T->value(temperature, Point());
+    Real e = _e_T->value(temperature, Point());
+
+    if (j > 0 && (e - e_old <= 0.0 || h - h_old <= 0.0))
+      mooseError("e(T) or h(T) are not monotonically increasing with T");
+    e_old = e;
+    h_old = h;
+  }
+
+  // backward lookup tables
+  _min_e = _e_T->value(_min_temperature, Point());
+  _max_e = _e_T->value(_max_temperature, Point());
+  _delta_e = (_max_e - _min_e) / ((Real)n - 1.0);
+  _min_h = _h_T->value(_min_temperature, Point());
+  _max_h = _h_T->value(_max_temperature, Point());
+  _delta_h = (_max_h - _min_h) / ((Real)n - 1.0);
+
+  _T_h_lookup.resize(n);
+  _T_e_lookup.resize(n);
+  for (unsigned int j = 0; j < n; ++j)
+  {
+    // internal energy
+    {
+      Real internal_energy = _min_e + j * _delta_e;
+
+      // The temperature is found by e - e(T) = 0
+      Real min = _min_temperature;
+      Real max = _max_temperature;
+      auto e_diff = [&internal_energy, this](Real x)
+      { return this->e_from_p_T(0.0, x) - internal_energy; };
+      BrentsMethod::bracket(e_diff, min, max);
+      Real temperature = BrentsMethod::root(e_diff, min, max);
+      _T_e_lookup[j] = temperature;
+    }
+
+    // enthalpy
+    {
+      Real enthalpy = _min_h + j * _delta_h;
+
+      // The temperature is found by e - e(T) = 0
+      Real min = _min_temperature;
+      Real max = _max_temperature;
+      auto h_diff = [&enthalpy, this](Real x) { return this->h_from_p_T(0.0, x) - enthalpy; };
+      BrentsMethod::bracket(h_diff, min, max);
+      Real temperature = BrentsMethod::root(h_diff, min, max);
+      _T_h_lookup[j] = temperature;
+    }
+  }
+}
+
+std::string
+CaloricImperfectGas::fluidName() const
+{
+  return "caloric_imperfect_gas";
+}
+
+Real
+CaloricImperfectGas::e_from_T(Real T) const
```

Comment:
Not in particular, but as long as we support non-AD in MOOSE, we should implement them, right?

Reply:
I just mean that it's possible that no one ever uses the non-AD version, so save the work until that point (and maybe the person who actually needs it can do the work!). But just an opinion - definitely ok to implement them.